File: Data Structures & Algorithms/buy-and-sell-crypto/submission-3.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        int buy = prices[0];

        for (int num : prices) {
            maxProfit = max(maxProfit, num - buy);
            if (num < buy) {
                buy = num;
            }
        }

        return maxProfit;
    }
};

File: Data Structures & Algorithms/maximum-subarray/submission-3.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # curMax = nums[0]
        # curSum = 0

        # for num in nums:
        #     if curSum < 0:
        #         curSum = 0

        #     curSum += num
        #     curMax = max(curSum, curMax)
        
        # return curMax

        curMax = nums[0]
        curSum = 0

        l = 0

        for r in range(len(nums)):
            if curSum < 0:
                l = r
                curSum = 0
            
            curSum += nums[r]
            curMax = max(curMax, curSum)

        return curMax